ENH: add max dist to NearestNDInterpolator

harshilkamdar opened the pull request:
#### Reference issue
This does not close a particular issue as far as I know, but does speed up nearest neighbor interpolation by a siginificant amount in cases where a max distance arg is passed.

#### What does this implement/fix?
This allows for far quicker queries and nearest neighbor interpolation for large datasets where prior information about when/where to interpolate is known. In some 2D geospatial examples with tens of millions of points, this leads to a speedup of anywhere from 2-30x. 

#### Additional information
n/a. This is my first PR here -- please let me know if I did anything wrong.

j-bowhay commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -89,13 +94,14 @@ class NearestNDInterpolator(NDInterpolatorBase):
     """
 
-    def __init__(self, x, y, rescale=False, tree_options=None):
+    def __init__(self, x, y, rescale=False, tree_options=None, query_max_dist=np.inf):
```
```suggestion
    def __init__(self, x, y, rescale=False, tree_options=None, *, query_max_dist=np.inf):
```

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -44,6 +44,11 @@ class NearestNDInterpolator(NDInterpolatorBase):
+        Option to truncate ``cKDTree`` nearest neighbor query to some
+        ``query_max_dist``. This is useful for larger interpolation problems.
+
+        .. versionadded:: 1.11.4
```
New features go to 1.12, I'd expect.

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,8 +122,16 @@ def __call__(self, *args):
         xi = self._scale_x(xi)
-        dist, i = self.tree.query(xi)
-        return self.values[i]
+        dist, i = self.tree.query(xi, distance_upper_bound=self.query_max_dist)
```
Since this new argument is forwarded to `kdtree.query`, maybe better keep the name consistent. Maybe even expose `**kwargs` and forward them all to `self.tree.query`. Not sure if they all make sense for the `NearestNDInterpolator`, so maybe start with `distance_upper_bound` and `workers`. Not sure about the rest, WDYT?

harshilkamdar replied:
Good catch on also exposing workers. The rest don't seem useful. Having **kwargs might be a little dangerous because the interpolator will break if k is not set to 1. I have exposed workers and distance_upper_bound (renamed) separately in the latest commit: https://github.com/scipy/scipy/pull/19483/commits/2fa01c9110a7c4db730625ac2efb57e94b8d7f3d. Let me know what you think

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,8 +122,16 @@ def __call__(self, *args):
-        return self.values[i]
+        dist, i = self.tree.query(xi, distance_upper_bound=self.query_max_dist)
+        if np.isfinite(self.query_max_dist):
+            # dist will be infinite for points where NN returned no neighbors. set those to nan, and mask in arr.
```
Is this what `kdtree` returns? Would be great to stress it in the comment if it's indeed the case.

harshilkamdar replied:
Added some more content in the text to stress this point and clarify.

ev-br reviewed (changes requested):
Looks very reasonable!

My main comment is about naming: I think it would be better to 
1) keep the name of the parameter consistent with KDTree, and 
2) maybe think about exposing other `KDTree.query` parameters where it makes sense. 

The second part here is optional of course.

ev-br commented:
One other thing (sorry for the piecemeal review, only noticed this being next to `**tree_options`): you are adding these new arguments to the constructor and only use them at the call site. Is there a need to store `workers` and `distance_upper_bound` on the instance? Maybe better mirror KDTree and only supply them in `__call__`.

harshilkamdar commented:
No worries at all. My rationale was looking to [`interpolate.griddata`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.griddata.html#scipy.interpolate.griddata) in the future and making it more flexible. 

Right now, griddata does not support passing any options to the underlying interpolator (nearest, linear, or cubic). I'm hoping to change that in a new PR and figured it would be cleaner to only expose and use class-level options and keeping __call__ clean for now. Linear doesn't have interesting options, but CloughTocher2DInterpolator [does](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CloughTocher2DInterpolator.html#scipy.interpolate.CloughTocher2DInterpolator). 

This is not a super strong reason, so I'm happy to go with whatever you prefer.

ev-br commented:
Right. So delineating between `__init__` and `__call__` optional arguments would require two sets of dicts as inputs to `griddata` (unpacked or not), and those would be `method`-dependent. Not nice. 
OTOH, you current approach is that everything is passed in a constructor and `__call__` signature only has data points and nothing else. Which is a reasonable invariant, so let's roll with it I'd say. WDYT @j-bowhay ?

I'd then ask for a small tweak:
- please add a leading underscore to things you store on the instance (as in `self._workers` is private, while `self.workers` is a part of public API).

In fact, it's not immediately clear from the docstring that this whole interpolator is for unstructured data and is KDTree-based. Would be great if you could update that in a follow-up PR.

j-bowhay commented:
I appreciate the argument around future work on `griddata` but it seems cumbersome that I would have to rebuild my interpolator just because I want to change the workers / max dist. Could we not add an `options` dict argument to `griddata` (as we do for minimise and friends) then each method uses this where needed (whether that be `__init__` or `__call__`)

harshilkamdar commented:
@j-bowhay, @ev-br: thank you both for your careful feedback. I have updated the PR based on both your comments in the latest commit - https://github.com/scipy/scipy/pull/19483/commits/ff1d52ffe1a827bf4941ccbc213dd5c21939f767. It now takes in a `query_options` dict and exposes some more cKDTree functionality. I've also added a couple more tests.

> In fact, it's not immediately clear from the docstring that this whole interpolator is for unstructured data and is KDTree-based. Would be great if you could update that in a follow-up PR.

I will do this in a separate documentation PR.

j-bowhay commented:
Sorry I wasn't quite clear, I was proposing the `griddata` take some kind of option dict to pass to the underlying interpolator and a signature like `NearestNDInterpolator.__call__(*args, distance_upper_bound=np.inf, etc)` but perhaps lets see what @ev-br thinks so we don't have to make lots of back and forth changes

ev-br commented:
To be clear: I'm fine with either approach. I'm sold on that the inconvenience of lumping together constructor and call arguments is rather minor and a cleaner `griddata` signature outweights it; I can also live with a `optimize`-like approach. As long as we are not adding public attributes to interpolators, I'm OK and defer to @j-bowhay for the final API specification.

harshilkamdar commented:
Sounds good - sorry about the added confusion. @j-bowhay let me know which option you prefer and I can implement. Either is fine on my end and should likely be minimal effort

j-bowhay commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -98,26 +98,55 @@ def __init__(self, x, y, rescale=False, tree_options=None):
         self.values = np.asarray(y)
 
-    def __call__(self, *args):
+    def __call__(self, *args, query_options=None):
```
```suggestion
    def __call__(self, *args, **query_options):
```

j-bowhay commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -98,26 +98,55 @@ def __init__(self, x, y, rescale=False, tree_options=None):
             or x1 can be array-like of float with shape ``(..., ndim)``
 
         """
+        if query_options is not None and not isinstance(query_options, dict):
+            raise TypeError("query_options must be a dictionary")
+
+        # gather the query options to pass to cKDTree.query
+        query_options = query_options or {}
+        eps = query_options.get('eps', 0)
+        p = query_options.get('p', 2)
+        distance_upper_bound = query_options.get('distance_upper_bound', np.inf)
+        workers = query_options.get('workers', 1)
+
```
```suggestion

```

j-bowhay commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -98,26 +98,55 @@ def __init__(self, x, y, rescale=False, tree_options=None):
         xi = self._scale_x(xi)
-        dist, i = self.tree.query(xi)
-        return self.values[i]
+        dist, i = self.tree.query(xi, eps=eps, p=p, distance_upper_bound=distance_upper_bound, workers=workers)
```
```suggestion
        dist, i = self.tree.query(xi, **query_options)
```

j-bowhay reviewed (commented):
How about an interface like this?

harshilkamdar commented:
Thanks both for your patience. @j-bowhay - I like it. I have committed and updated tests + docs and have re-requested review.

ev-br commented:
CI failures look related, could you please take a look

harshilkamdar commented:
Sorry, dumb bug. Fixed now.

ev-br commented:
CI is still unhappy :-(.
Might be faster to iterate locally: the incantation is `$ python dev.py test -s interpolate` or `-t path/to/test/file`.

harshilkamdar commented:
@ev-br - good call. 

Through the `griddata` test failures, I've found that both `NearestNDInterpolator` and `griddata` are silently actually supporting two things that I wasn't originally accounting for - the case where `__call__` gets an n-D `xi` argument and the case where the y-values for the interpolation are also multidimensional. 

I've come up with hopefully an acceptable solution to account for these two cases and updated docs & comments for NearestNDInterpolator to reflect this. I couldn't get scipy to build locally for some annoying reasons but have tried to make sure that all tests in `test_ndgriddata.py` pass by checking manually. If there end up being failures, I can fix them quickly.

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -30,9 +30,9 @@ class NearestNDInterpolator(NDInterpolatorBase):
+    x : (npoints, ndims) n-D ndarray of floats
         Data point coordinates.
-    y : (npoints, ) 1-D ndarray of float or complex
+    y : (npoints, ...) n-D ndarray of floats or complex
```
OK, so the original was papering over a detail that `y` can have trailing dimensions (BTW, I'm not sure LinearNDInterpolator and CT can). 
The new version is plain confusing, my mind starts racing to figure out what is `n` in  `n-D`. Essentially, what is going on here is that `y` is a 1D array of data values or a stack of 1D arrays of data values, correct? I'd probably say something along these lines then (ML people like to talk about batch dimensions).

harshilkamdar replied:
I realized my response to this didn't go through. Yeah, you're right - I've fixed this in latest commit. Looks like Linear and CT can actually. There's some voodoo going on in NDInterpolatorBase's _set_values, where it'll recast values to be 2D (npoints, nvalues) for all interoplators that subclass.

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -30,9 +30,9 @@ class NearestNDInterpolator(NDInterpolatorBase):
     Parameters
     ----------
-    x : (npoints, ndims) 2-D ndarray of floats
+    x : (npoints, ndims) n-D ndarray of floats
```
`x` is surely a 2D array, no trailing dimensions allowed.

ev-br reviewed (commented):
OK, this is getting close! I like the flatten-compute-restore logic, it makes a lot of sense. 

I've left several comments about 1) minor simplifications of implementation details and 2) clarity of comments/documentation.

The story about stacks of arrays is always confusing, so let's try to clarify it here. So, there are three arrays: `x`, `y` and `xi`. 

Let's see if I got it right:

- `x` is always 2D
- `y` is 1D or a stack of 1D, where the first dimension is the same as the second (or first?) dimension of `x`, and trailing dimensions represent the stack of 1D arrays.
- `xi` is normally a 1D array of dimension `ndim` or a stack of 1D arrays, where the stacking is along the _leading_ dimension (this is what _broadcastable_ word hints at inhttps://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.NearestNDInterpolator.__call__.html#scipy.interpolate.NearestNDInterpolator.__call__ ).

I think it would be best to explain this (or a correct version of it) to clearly separate what is the number of dimensions, what is the number of data points and how stacked dimensions come into play.
Two general considerations to keep in mind:

- the docs for NearestNDInterpolator, LinearNDInterpolator and CloughtTocher2DInterpolator should look relatively consistent. If one of them supports trailing/leading dimensions and the other does not, OK, so be it. But going from nearest to linear interpolator in a simple case should be seamless, and so should be the docs.
- Maybe it'd be useful to take a look at how RegularGridInterpolator documents essentially the same issue. Ideally, regular and scattered interpolants look similar.

Finally, since the documentation story is not strictly speaking related to your original changes, it would be perfectly fine to undo the docstring changes and keep the documentation update for a follow-up PR (which would be very welcome indeed).

j-bowhay commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -108,6 +108,11 @@ def __call__(self, *args):
             Points where to interpolate data at.
             x1, x2, ... xn can be array-like of float with broadcastable shape.
             or x1 can be array-like of float with shape ``(..., ndim)``
+        query_options : dict, optional
```
This 
```suggestion
        **query_options
```

harshilkamdar commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,14 +121,50 @@ def __call__(self, *args):
+        if self.values.ndim == 1:
+            interp_values[valid_mask] = self.values[i[valid_mask]]
+        else:
+            interp_values[valid_mask] = self.values[i[valid_mask], ...]
```
To clarify - is what I've wrriten right now adequate or should I modify?

ev-br replied:
I would be tempted to just write `interp_values[valid_mask] = self.values[i[valid_mask], ...]` but it's OK, can do in a sequel. No need to push a fix just for this one.

harshilkamdar replied:
Suonds good - will do

harshilkamdar commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,14 +121,50 @@ def __call__(self, *args):
+        if np.issubdtype(self.values.dtype, np.complexfloating):
+            interp_values = np.full(interp_shape, np.nan, dtype=self.values.dtype)
+        else:
+            interp_values = np.full(interp_shape, np.nan)
```
Exactly. There were a few tests where self.values were integers and it caused numpy weirdness when I tried to assign nan's to an array declared as int. example below 


```
In [1]: arr = np.full(5, np.nan, dtype=np.int64)

In [2]: arr
Out[2]: 
array([-9223372036854775808, -9223372036854775808, -9223372036854775808,
       -9223372036854775808, -9223372036854775808])

```

harshilkamdar commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,14 +121,50 @@ def __call__(self, *args):
+        # (1) the case where xi is of some dimension (n, m, ..., D), where D is the coordinate dimension, and
+        # (2) the case where y is multidimensional (npoints, k, l, ...).
+        # We will first flatten xi to deal with case (1) and build an intermediate return array with shape
+        # (n*m*.., k, l, ...) and then reshape back to (n, m, ..., k, l, ...).
```
Yeah, I'm not being careful here in my words. Have given the description another go.

harshilkamdar commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -116,14 +121,50 @@ def __call__(self, *args):
-        dist, i = self.tree.query(xi)
-        return self.values[i]
+
+        # We need to handle two important cases for compatibility with a flexible griddata:
```
Yeah, this is confusing and griddata is not relevant - updated.

harshilkamdar commented:
@ev-br - thanks for the thorough comments. I think you're right that I'm confusing nD/2D and trailing dimensions. The pointer to RegularGridInterpolator was very helpful.

I have considered most of your comments and hopefully addressed bulk of it in the new commit. Hopefully this is a more accurate version of what's going on.

ev-br commented on scipy/interpolate/_ndgriddata.py:
```diff
@@ -108,6 +108,12 @@ def __call__(self, *args):
+            function to be explicitly set. See the `scipy.spatial.cKDTree.query` for an overview of the different
+            options.
+
+            ..versionadded:: 1.12.0
```
```suggestion
            .. versionadded:: 1.12.0
```

ev-br reviewed (approved):
All my comments have been addressed, I think it is a great addition, so +1 from me. 

Unless there are further comments, am going to hit the green button later this week.

ev-br commented:
Am going to merge now, based on two core dev approvals. Thanks Jake for the review, thank you @harshilkamdar for the enhancement and congrats with what I believe is your first SciPy commit. Keep them coming!
